Remove usage of goog.color and closure named colors

bjornharrtell opened the pull request:
There are some mentions of goog.color in the comments, perhaps those should be removed too? In that case the private member ol.color.rgbColorRe_ will have no explanation, is that ok?

tschaub commented:
> There are some mentions of goog.color in the comments, perhaps those should be removed too

Yes, these comments should be updated/removed.

In addition, we'll need to add upgrade notes for people who were using `ol.ENABLE_NAMED_COLORS`.

sweco-sebhar commented:
Amended with updated comments and upgrade note.

tschaub commented on src/ol/color/color.js:
```diff
@@ -208,11 +201,6 @@ ol.color.fromString = (
-    // goog.color.names does not have a type declaration, so add a typecast
-    s = /** @type {string} */ (goog.color.names[s]);
-    isHex = true;
-  }
```
It looks like `isHex` should be removed entirely now.

marcjansen commented on changelog/upgrade-notes.md:
```diff
@@ -6,6 +6,11 @@
 
+#### Removal of `ol.ENABLE_NAMED_COLORS`
+
+This option existed to support the use of Google Closure library named colors and was removed as a part of OpenLayers moving away from Google Closure library. Note that this do not affect the use of CSS named colors.
```
Typo: ... this do**es** not ...

marcjansen replied:
Should we more explicit here? Sth. like: if you have used the canvas renderer (which is the default) for your vector layers, this change will not affect you. If you have been using the webgl renderer though, you cannot any longer use named colors but must provide them in <a suitable format.

tschaub replied:
I agree it would be nice to be more explicit about what this change means.  I would remove all reference to the Google Closure Library, since I don't think that is relevant.  It would be more relevant for people to know if they need to change their application code.

sweco-sebhar replied:
How about:

Note that this option default to `false` and that it has no effect on the canvas renderer (which is the default). If you have relied on this option to use named colors (created with `ol.color.fromString`) for use with the WebGL renderer you will need to translate them to RGB values.

bjornharrtell commented:
Thanks for the input, updated branch with isHex removed and new update notes based on feedback.

gberaudo commented:
This removal will affect our ability to handle named colors in the WEBGL renderer.
To avoid this, a straightforward solution is to copy the [goog color table](
https://closure-library.googlecode.com/git-history/docs/local_closure_goog_color_names.js.source.html#line28) into our code.

bjornharrtell commented:
Sure, could do that. I was actually initially going that way but decided/thought it wasn't a feature worth keeping. Opinions from previous reviewers @tschaub and @marcjansen?

marcjansen commented:
Fine for me if it isn't included if WEBGL isn't included.

sweco-sebhar commented:
I've made so that this PR only removes the use of goog.color. Cannot change the title of the PR to reflect that `ol.ENABLE_NAMED_COLORS` is left instact, but the commit message should be ok.

bjornharrtell commented:
I've made so that this PR only removes the use of goog.color. Cannot change the title of the PR to reflect that ol.ENABLE_NAMED_COLORS is left instact, but the commit message should be ok.

gberaudo commented:
Looks great to me. Thank you @bjornharrtell.

ahocevar commented:
I really think we don't need named colors in our library. If someone wants to use named colors with WegGL, this can be translated with something like
```js
function getRgbColor(color) {
  var el = document.createElement("div");
  el.style.color = "white";
  document.body.appendChild(el);
  var rgb = window.getComputedStyle(el).color;
  document.body.removeChild(el);
};
```
This could be done in the WebGL renderer, and of course `el` would not have to be created, added to the DOM and removed with every single call.

bjornharrtell commented:
I agree. I'll wait for consensus before having another go at this PR.

ahocevar commented:
Sorry @bjornharrtell for the back and forth on this.

gberaudo commented:
@ahocevar, using `window.getComputedStyle` sounds like a great idea.

Having the conversion in `ol.color.fromStringInternal_` would notably allow to share the cache; I think it would be better.

It would also make sense to have `ol.ENABLE_NAMED_COLORS` enabled by default for WEBGL users and allow to disable it as an optimization.

ahocevar commented:
> Having the conversion in ol.color.fromStringInternal_ would notably allow to share the cache; I think it would be better.

Fine with me.

> It would also make sense to have ol.ENABLE_NAMED_COLORS enabled by default for WEBGL users and allow to disable it as an optimization.

We can then get rid of ol.ENABLE_NAMED_COLORS, because we don't keep all the color names in the code any more.

gberaudo commented:
> We can then get rid of ol.ENABLE_NAMED_COLORS, because we don't keep all the color names in the code any more.

OK, it is fine with me if we remove it.

probins commented:
what's the status with this? As NAMED_COLORS is disabled by default, and AFAIK not documented anywhere, I doubt whether many people even know of its existence.

bjornharrtell commented:
I've rebased and made the changes so this PR simply removes `NAMED_COLORS` and any use of `goog.colors`.

I'm not sure how to complete the implementation suggestion using `window.getComputedStyle`, in particular to make it conditional in a way that does not slow or disrupt the Canvas renderer, so if it's still desirable to support named colors with the WebGL renderer that remains to be done.

fredj commented:
you can remove tests in `test/spec/ol/color.test.js`

gberaudo commented:
@bjornharrtell, if it is too complex to implement, then could you please update the `changelog/upgrade-notes.md` to state that using named color is no longer possible with the `WEBGL` renderer?

bjornharrtell commented:
Thanks @gberaudo and @probins. I've changed the upgrade notes to explain that named colors are no longer supported with WebGL. I've removed one test and adapted the other that tests the cache since it still seem relevant.

I took another look at the suggested method from @ahocevar and while it looks like a neat hack I still do not see how to properly integrate it.

ahocevar commented:
@bjornharrtell There were two typos in my [snippet](https://github.com/openlayers/ol3/pull/4194#issuecomment-146452741). I just updated it.

probins commented on changelog/upgrade-notes.md:
```diff
@@ -4,6 +4,10 @@
 
+#### Removal of `ol.ENABLE_NAMED_COLORS`
+
+This change only affects the WebGL renderer and allowed use of named colors for that renderer. Note that the default setting was false and that the Canvas and DOM renderers allows named colors regardless.
```
... renderers allow ... (just noticed this)

probins commented:
I'm a bit confused as to who should do what so this can be merged.

bjornharrtell commented:
As the WebGL renderer is progressing, I'd like to take a second try at implementing the proposal from @ahocevar.

bjornharrtell commented:
I've reimplemented so that named colors should be translated and cached. However I'm not sure how named colors are supported currently with `ol.ENABLE_NAMED_COLORS` at the default `false` and the canvas renderer - I suppose they don't need translation but I can't see the code path allowing that.

ahocevar commented:
Very nice, thanks @bjornharrtell. With such a small code snipped, it was definitely the right decision to get rid of `ol.ENABLE_NAMED_COLORS`.

probins commented:
@bjornharrtell see https://github.com/openlayers/ol3/pull/1985#issuecomment-50462922 `goog.color.names` are irrelevant for Canvas, which uses names like 'cyan' etc whether `ENABLE_NAMED_COLORS` is set or not

bjornharrtell commented:
Yeah I get that but I was worrying that my changes would mean that named colors now will be translated for the Canvas renderer regardless.

probins commented:
unnecessary added overhead, you mean. Also, is it necessary to create/add a new DOM element for every name? Would it not be better to have one named element and set the color?

ahocevar commented:
As far as I can tell, `ol.color.asArray` is only used by `ol.style.Icon` when the `color` option is set  and `ol.format.KML` when writing KML. I don't see any renderer related use.

ahocevar commented:
So the place to improve things is `ol.style.Icon`, where the color could be taken from the options as-is. Only the webgl renderer would have to call `ol.color.asArray`.

Any takers for a pull request with this improvement?
